Revise "Data Storage and Migration" and add Examples for Migration via Hooks (#1917)
rauc/rauc: docs/advanced.rst

Data Storage and Migration
--------------------------

Most systems require one or multiple locations for storing persistent data such
as application configuration, user settings or runtime state.
When performing an update, you have to ensure that the updated system can
take over or access the data from the previous installation.

Persistent data can be stored in different ways:

* directly inside the root file system,
* on separate shared data partitions, or
* on redundant data partitions.

.. image:: images/data_migration.svg
  :width: 600
  :align: center

Each approach has different implications for update and fallback handling,
as discussed in the following sections.

Storing Data in The Root File System
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In a typical Linux system, configuration data is stored under standard paths
such as ``/etc``, which are part of the root file system by default.
If the root file system is writable, it can directly hold persistent data.
When performing the update, you have to ensure that the files you need to
preserve are copied to the target slot after having written
the system data to it.

> See :ref:`sec-data-migration-rootfs` for information on migration rootfs data.

Reviewer: ```suggestion
See :ref:`sec-data-migration-rootfs` for information on migrating data in the rootfs.
```

It depends on the amount and type of data you want to handle which option you
should choose.

Managing a ``/dev/data`` Symbolic Link
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For redundant data partitions the active rootfs slot has to mount the correct
data partition dynamically.

> For example with ubifs, a udev rule set can be used for this::

Reviewer: Shouldn't we recommend the symlinks RAUC creates in `/run/rauc` for that? Can be a separate PR though.

Author: I've added a commit addressing this.

| Fallback  | tricky (reconvert data?) | easy (old data!)          |
+-----------+--------------------------+---------------------------+

While RAUC does not handle data migration directly, it provides hooks that
can be used to implement it in the bundle (or handlers for a system-defined
handling).
Note that in many cases, migration of application data is best handled by
the application itself on first boot after the update.

.. _sec-data-migration-rootfs:

Migrating Root File System Data
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For migrating data from the current rootfs to the updated rootfs,
you can use a *post-install* :ref:`slot hook <sec-slot-hooks>` for your rootfs
slot class.

.. code-block:: cfg

   [image.data]
   filename=rootfs.img
   hooks=post-install

> .. note:: This requires a writeable rootfs slot (type).

Reviewer: ```suggestion
.. note:: This requires a rootfs slot type which can be mounted read-write (e.g. ext4).
```

The hook will run after the new rootfs image has been written to the target

> slot, with the target slot still mounted.

Reviewer: ```suggestion
slot, with the target slot mounted.
```

When using images for ext4 slots, it will be mounted after writing the image, so "still" is not really correct.

This allows copying relevant files from the currently active rootfs into the
new slot before the system reboots into it.

A hook script could look as follows:

.. code-block:: sh

   #!/bin/sh

   case "$1" in
           slot-post-install)
                   # only data should be handled
                   test "$RAUC_SLOT_CLASS" = "rootfs" || exit 1

                   mkdir -p "$RAUC_SLOT_MOUNT_POINT/etc/"

                   echo "Preserving /etc/timezone"
                   cp -a /etc/timezone "$RAUC_SLOT_MOUNT_POINT/etc/"
                   echo "Preserving /etc/passwd and /etc/shadow"
                   cp -a /etc/passwd "$RAUC_SLOT_MOUNT_POINT/etc/"
                   cp -a /etc/shadow "$RAUC_SLOT_MOUNT_POINT/etc/"
                   ;;
           *)
                   exit 1
                   ;;
   esac

>    exit 0

Reviewer: Add a note that this is focused on the basic functionality and integration with RAUC, and that error handling was intentionally left out.

Author: I switched to make it less look like a comprehensive example instead with adding ellipsis and noting 'simplified shell script code'. This also aligns better with existing examples.

Migration for Redundant Data Partitions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Migrating data between redundant data partitions is quite specific.
Thus RAUC does not provide any built-in mechanism for this,
but it provides the necessary primitives to implement it via hooks.

You'll need to describe the partitions as slots with a mountable file system:

.. code-block:: cfg

   [slot.data.0]
   type=ext4
   ...

   [slot.data.1]
   type=ext4
   ...

The manifest should then specify ``type=empty-fs``,
which will make RAUC format the target slot depending on its slot type
(only ext4 is supported for now).
For the actual data copying or migration, a ``post-install`` hook must be used,

> which will be called with the target slot already mounted at ``$RAUC_SLOT_MOUNT_POINT``:

Reviewer: ```suggestion
which will be called with the target slot mounted at ``$RAUC_SLOT_MOUNT_POINT``:
```

> 

Reviewer: Perhaps add: "Conceptually, RAUC behaves just as if the bundle contains an empty tar archive for this slot."

.. code-block:: cfg

   [image.data]
   type=empty-fs
   hooks=post-install

An example hook could look as follows:

.. code-block:: sh

   #!/bin/sh

   case "$1" in
           slot-post-install)

>                    # only data should be handled

Reviewer: ```suggestion
                   # only the data slots should be handled
```

                   test "$RAUC_SLOT_CLASS" = "data" || exit 1

                   echo "Migrating data from /data to $RAUC_SLOT_MOUNT_POINT"
                   rsync -a --delete /data/ "$RAUC_SLOT_MOUNT_POINT/"
                   ;;
           *)
                   exit 1
                   ;;
   esac

   exit 0

.. note::

   This example assumes the currently active data partition is mounted at ``/data``.
   The ``--delete`` flag ensures the target slot is an exact mirror of the source,

>    removing any stale files left from a previous installation.

Reviewer: Also add a note about error handling.

Author: See above.

+-----------+--------------------------+---------------------------+

While RAUC does not handle data migration directly, it provides hooks that
can be used to implement it in the bundle (or handlers for a system-defined
handling).
Note that in many cases, migration of application data is best handled by
the application itself on first boot after the update.

.. _sec-data-migration-rootfs:

Migrating Root File System Data
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For migrating data from the current rootfs to the updated rootfs,
you can use a *post-install* :ref:`slot hook <sec-slot-hooks>` for your rootfs
slot class.

.. code-block:: cfg

   [image.data]
   filename=rootfs.img
   hooks=post-install

.. note:: This requires a rootfs slot type which can be mounted read-write
   (e.g. ext4).

The hook will run after the new rootfs image has been written to the target
slot, with the target slot mounted.
This allows copying relevant files from the currently active rootfs into the
new slot before the system reboots into it.

A simplified hook shell script code could look as follows:

.. code-block:: sh

   case "$1" in

           [...]

           slot-post-install)

>                    # only the data slots should be handled

Reviewer: ```suggestion
                   # only the rootfs slots should be handled
```

   exit 0

Migration for Redundant Data Partitions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Migrating data between redundant data partitions is quite specific.

Reviewer: Specific to what? Perhaps drop this sentence?

Author: Without this, also the following sentence wouldn't make sense. Thus I tried to be more specific about what's specific here instead.
